src/Agent.Listener/Configuration/WindowsSecurityManager.cs

```diff
@@ -0,0 +1,477 @@
+        {
+            Trace.Info(nameof(GrantFullControlToGroup));
+            DirectoryInfo dInfo = new DirectoryInfo(path);
+            DirectorySecurity dSecurity = dInfo.GetAccessControl();
+
+            if (!dSecurity.AreAccessRulesCanonical)
+            {
+                Trace.Warning("Acls are not canonical, this may cause failure");
+            }
+
+            dSecurity.AddAccessRule(
+                new FileSystemAccessRule(
+                    groupName,
+                    FileSystemRights.FullControl,
+                    InheritanceFlags.ContainerInherit | InheritanceFlags.ObjectInherit,
+                    PropagationFlags.None,
+                    AccessControlType.Allow));
+            dInfo.SetAccessControl(dSecurity);
+        }
+
+        public bool GrantUserLogonAsServicePrivilage(string accountName)
+        {
+            Trace.Info(nameof(GrantUserLogonAsServicePrivilage));
+
+            IntPtr lsaPolicyHandle = IntPtr.Zero;
+            string domain = string.Empty;
+            string user = string.Empty;
+            GetAccountSegments(accountName, out domain, out user);
+
+            try
+            {
+                LSA_UNICODE_STRING system = new LSA_UNICODE_STRING();
+
+                LSA_OBJECT_ATTRIBUTES attrib = new LSA_OBJECT_ATTRIBUTES()
+                {
+                    Length = 0,
+                    RootDirectory = IntPtr.Zero,
+                    Attributes = 0,
+                    SecurityDescriptor = IntPtr.Zero,
+                    SecurityQualityOfService = IntPtr.Zero,
+                };
+
+                uint result = LsaOpenPolicy(ref system, ref attrib, LSA_POLICY_ALL_ACCESS, out lsaPolicyHandle);
+                if (result != 0 || lsaPolicyHandle == IntPtr.Zero)
+                {
+                    throw new Exception(StringUtil.Format("Open Lsa failed: hr={0}", result));
+                }
+
+                result =LsaAddAccountRights(lsaPolicyHandle, GetSidBinaryFromWindows(domain, user), LogonAsServiceRights, 1);
+                Trace.Info("LsaAddAccountRights return with error code {0} ", result);
+
+                return result == 0;
+            }
+            finally
+            {
+                var result = LsaClose(lsaPolicyHandle);
+                if (result != 0)
+                {
+                    Trace.Error(StringUtil.Format("Can not close LasPolicy handler. LasClose failed with error code {0}", result));
+                }
+            }
+        }
+
+        public static void GetAccountSegments(string account, out string domain, out string user)
+        {
+            string[] segments = account.Split('\\');
+            domain = string.Empty;
+            user = account;
+            if (segments.Length == 2)
+            {
+                domain = segments[0];
+                user = segments[1];
+            }
+        }
+
+        public bool IsValidCredential(string logonAccount, string logonPassword)
```

Comment:
It would be really cool if there is an easy way to translate the result into a useful error message. For example if their account is locked, then there is no use retrying.